### DISCUSSION TITLE
How to spy on inline declared functions like gtag?

### DISCUSSION THREAD

**Discussion by alewolf** [cypress-io/cypress#15694] [Questions and Help] [answered]
I want to spy on the Google Analytics `gtag` calls right after the `gtag` has been initialized. All my attemtps to spy on `gtag` with Cypress fail. Either I am too early with hooking in and get the error `Attempted to wrap undefined property gtag as function`, or I am too late with hooking in and the spy doesn't register any calls at all. 

Mind you I do have working tests with the spy. But, they all capture `gtag`calls on button clicks for example, which happen much later in the flow. 

Example HTML that sets up `gtag` and then fires a `gtag` event that I want to spy on: 

```html
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-12345678-9"></script>

<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() {
        dataLayer.push(arguments);
    }
    gtag('js', new Date());
    gtag('config', 'UA-12345678-9');
    gtag('event', 'view_item', {
        "send_to": 'UA-39746956-9',
        "items": []
    });
</script>
```

I want to spy on the `gtag('event', 'view_item', {...})` event.

When trying to set up the spy with `onBeforeLoad` I get the `Attempted to wrap undefined property gtag as function` error. 

```js
it('visit product page', () => {
    cy.visit('/product/shirt/', {
        onBeforeLoad(win) {
            cy.spy(win, 'gtag').as('gtag')
        },
    })
})
```

On the other hand, when I try to wait until `gtag` is loaded and then try to set up the spy, it's too late and never catches the `gtag` calls. 

```js
it('visit product page', () => {
    cy.visit('/product/shirt/')
    cy.window().its('gtag')
    cy.window().then((win) => {
        cy.spy(win, 'gtag').as('gtag');
    })
})
```

Or the following code returns the error `Identifier gtag has already been declared`.

```js
it('visit product page', () => {
    const gtag = cy.stub().as('gtag')
    cy.on('window:before:load', (win) => {
        Object.defineProperty(win, 'gtag', {
            configurable: false,
            get: () => gtag,
            set: () => { },
        })
    })
    cy.visit('/product/shirt/')
})
```

How can I set this up correctly, in order to spy on those `gtag` calls?

**Comment by bahmutov** [comment]
Can you try removing `configurable: false` from your `Object.defineProperty`? Other than that, similar code is working in

- https://glebbahmutov.com/blog/stub-window-open/
- see "Stubbing" slides in https://github.com/cypress-io/testing-workshop-cypress#advanced

If it still does not work, do you mind create a small repo with a reproducible example we could run?

**Reply by alewolf** [reply]
Removing `configurable: false` didn't work. 

Also, going through the documentation you've sent didn't help. I went through a few iterations in my code, but no luck. 

Therefore I've set up a repo as you suggested: https://github.com/wblabio/cypress-spy

**Comment by bahmutov** [comment, marked as answer]
I see you are using some custom local domain for testing, can you update your code so I can run it? Maybe add a local server and use a localhost:3000 for simplicity?

<img width="797" alt="Screen Shot 2021-04-06 at 11 35 49 AM" src="https://user-images.githubusercontent.com/2212006/113737676-3fa1df80-96cc-11eb-81f1-46ac4f970506.png">

**Reply by alewolf** [reply]
Ok. I quickly added a local npm server to the repo. 

You can spin it up by running `npx http-server` from within the repo. 

And then `npx cypress open` for the tests.

**Reply by bahmutov** [reply]
Ok, looked at it. Unfortunately you cannot use global variable definition `function gtag` - it won't be stubbable and will clash with second definition if you try to do `Object.defineProperty(win, 'gtag')` - just like you cannot declare the same variable multiple times.

Thus instead of this in your code

```js
function gtag() {
    dataLayer.push(arguments);
}
```

You should use this:

```js
window.gtag = function gtag() {
    dataLayer.push(arguments);
}
```

and then stub for each window load

```js
it('visit index page 4', () => {
  const gtag = cy.stub().as('gtag')
  cy.on('window:before:load', (win) => {
      console.log('win before load')
      Object.defineProperty(win, 'gtag', {
          get: () => gtag,
          set: () => {
              console.log('setting gtag')
          },
      })
  })
  cy.visit('/index.html')

  cy.get('@gtag').should('be.called')
  cy.get('@gtag').should('be.called', 'event', 'view_item')
})
```

**Reply by alewolf** [reply]
Ok. Thanks for that solution!

I don't understand what the difference is between

```js
function gtag() {
    dataLayer.push(arguments);
}
```

and 

```js
window.gtag = function gtag() {
    dataLayer.push(arguments);
}
```

It does make the test work though. 

More importantly. Do you see any risk changing that code in my entire application to:

```js
window.gtag = function gtag() {
    dataLayer.push(arguments);
}
```

It should work the same, shouldn't it?

**Reply by bahmutov** [reply]
I have not investigate it fully, but I believe the `function gtag` declares a new variable, so just like you cannot do `var gtag; var gtag` you cannot do `function gtag; Object.defineProperty(window, 'gtag')` in the test. There should be no problem using the property on the window

**Reply by alewolf** [reply]
I've run into two issues trying to extend my tests on this repo. 

1. I can't print the captured `@gtag` data to the console. I get a `Cannot read property 'name' of null` error.  This works when I use spies. But in this case I cannot use a spy right?

![image](https://user-images.githubusercontent.com/7146794/114497626-d9184680-9c22-11eb-8db3-3ec039aa7d1a.png)

2. I've added one more test to assert if a certain parameter was sent. 

`cy.get('@gtag').should('be.calledWith', 'event', 'purchase', Cypress.sinon.match.has("send_to", "UA-12345678-1"))`

Again, this works fine when I use a spy for events that happen after page load. But in this case I need to use a stub, which somehow works differently, or has a bug.

**Reply by bahmutov** [reply]
The first issue is https://github.com/cypress-io/cypress/issues/15816

The second issue - can you update the repo and provide a reproducible example showing the problem?

**Reply by alewolf** [reply]
I updated the repo with that new assertion: https://github.com/wblabio/cypress-spy

Let me know if you need additionally an example where it works fine with a spy.

**Comment by bahmutov** [comment]
I don't see a problem in your code, except the app and the test do not work (for both cy.spy and cy.stub cases)

cy.spy fails

<img width="716" alt="Screen Shot 2021-04-15 at 4 06 38 PM" src="https://user-images.githubusercontent.com/2212006/114931884-d4ab8380-9e04-11eb-8346-a2b21264d9db.png">

cy.stub fails

<img width="707" alt="Screen Shot 2021-04-15 at 4 07 00 PM" src="https://user-images.githubusercontent.com/2212006/114931920-e1c87280-9e04-11eb-9efc-385b95db2003.png">

in both cases, the calls are "event", "view" and never the "event", "purchase", so please check your test

<img width="744" alt="Screen Shot 2021-04-15 at 4 07 59 PM" src="https://user-images.githubusercontent.com/2212006/114931971-f4db4280-9e04-11eb-8613-0a1d931bb669.png">
